src/libs/print_settings.c

```diff
@@ -280,101 +271,186 @@ _print_button_clicked (GtkWidget *widget, gpointer user_data)
-  dt_pdf_finish(pdf, &pdf_page, 1);
+    params->pdf_image->rotate_to_fit = FALSE;
 
-  // free memory
+  params->pdf_page = dt_pdf_add_page(pdf, &params->pdf_image, 1);
+  dt_pdf_finish(pdf, &params->pdf_page, 1);
 
-  free (dat.ps->buf);
-  free (pdf_image);
-  free (pdf_page);
+  if(dt_control_job_get_state(job) == DT_JOB_STATE_CANCELLED) return 0;
+  dt_control_job_set_progress(job, 0.95);
 
   // send to CUPS
 
-  dt_print_file (imgid, filename, &ps->prt);
-
-  g_unlink(filename);
+  dt_print_file (params->imgid, params->pdf_filename, params->job_title, &params->prt);
+  dt_control_job_set_progress(job, 1.0);
 
   // add tag for this image
 
   char tag[256] = { 0 };
   guint tagid = 0;
-  snprintf (tag, sizeof(tag), "darktable|printed|%s", ps->prt.printer.name);
+  snprintf (tag, sizeof(tag), "darktable|printed|%s", params->prt.printer.name);
   dt_tag_new(tag, &tagid);
-  dt_tag_attach(tagid, imgid);
+  dt_tag_attach(tagid, params->imgid);
+
+  return 0;
+}
+
+static void _print_job_cleanup(void *p)
+{
+  dt_lib_print_job_t *params = p;
+  if(params->pdf_filename[0]) g_unlink(params->pdf_filename);
+  free(params->pdf_image);
+  free(params->pdf_page);
+  free(params->buf);
+  g_free(params->style);
+  g_free(params->buf_icc_profile);
+  g_free(params->p_icc_profile);
+  g_free(params->job_title);
+  free(params);
+}
+
+static void
+_print_button_clicked (GtkWidget *widget, gpointer user_data)
+{
+  const dt_lib_module_t *self = (dt_lib_module_t *)user_data;
+  dt_lib_print_settings_t *ps = (dt_lib_print_settings_t *)self->data;
+
+  const int imgid = dt_view_filmstrip_get_activated_imgid(darktable.view_manager);
+
+  if (imgid == -1)
+  {
+    dt_control_log(_("cannot print until a picture is selected"));
+    return;
+  }
+  if (strlen(ps->prt.printer.name) == 0 || ps->prt.printer.resolution == 0)
+  {
+    dt_control_log(_("cannot print until a printer is selected"));
+    return;
+  }
+  if (ps->prt.paper.width == 0 || ps->prt.paper.height == 0)
+  {
+    dt_control_log(_("cannot print until a paper is selected"));
+    return;
+  }
+
+  dt_job_t *job = dt_control_job_create(&_print_job_run, "print image %d", imgid);
+  if(!job) return;
+
+  dt_lib_print_job_t *params = calloc(1, sizeof(dt_lib_print_job_t));
+  dt_control_job_set_params(job, params, _print_job_cleanup);
+
+  params->imgid = imgid;
+  memcpy(&params->prt, &ps->prt, sizeof(dt_print_info_t));
+
+  // what to call the image?
+  GList *res;
+  if((res = dt_metadata_get(params->imgid, "Xmp.dc.title", NULL)) != NULL)
+  {
+    // FIXME: in metadata_view.c, non-printables are filtered, should we do this here?
+    params->job_title = g_strdup((gchar *)res->data);
+    g_list_free_full(res, &g_free);
+  }
+  else
+  {
+    const dt_image_t *img = dt_image_cache_get(darktable.image_cache, params->imgid, 'r');
+    if(!img || img->film_id == -1)
+    {
+      dt_control_log(_("cannot get image %d for printing"), imgid);
+      dt_control_job_dispose(job);
+      return;
```

Comment:
Shouldn't you release `img` first?

Reply:
Oh, good point. I'll fix it. I cargo-cult style check for `(!img || img->film_id == -1)`, but is it enough to just check for `!img`? In which case `dt_image_cache_read_release()` is a NOP if there is no image.

Reply:
You are right, film_id = -1 is probably a thing of the past when we still had the special single image film roll.

Reply:
OK! One more commit, to switch it back, with an explanatory comment why.